### PR TITLE
[Reviewer: Rowena] Allow LZ4 alongside zlib

### PR DESCRIPTION
**Pull request by rkd-msw**
Rowena, can you please review this change from a SAS team perspective? It's already been reviewed within the Clearwater team.

You're also reviewing my design, which I've sent by email - let me know if you have any design feedback which I should apply before you do the code review.

[Previous comment here, to Matt, was: "Heavily based on a) the LZ4 code you wrote and b) the design you've reviewed, so should be pretty straightforward. I've confirmed that this compiles but haven't live-tested yet."]

### CODE REVIEW

File: include/sas.h [hunk 1]
```diff
@@ -114,42 +110,46 @@ class SAS
...
   {
   public:
-    inline Profile(std::string dictionary) : _dictionary(dictionary) {}
+    enum CompressionType {
+      ZLIB = 0,
+      LZ4
+    };
+
+    inline Profile(std::string dictionary, CompressionType t = ZLIB) : _dictionary(dictionary), _type(t) {}
+    inline Profile(CompressionType t) : _dictionary(""), _type(t) {}
```
Comment (mirw) [inline include/sas.h]: I was going to say that you can use a delegating constructor here, i.e. `Profile(CompressionType t) : Profile("", t) {}`.  However, I think this is C++ 11 only, and I'm not sure whether our sas-client assumes that.  (Clearwater does.)

File: include/sas.h [hunk 2]
```diff
@@ -114,42 +110,46 @@ class SAS
...
     inline const std::string& get_dictionary() const {return _dictionary;}
+    inline bool is_lz4() const {return _type == LZ4;}
```
Comment (mirw) [inline include/sas.h]: Why not a `getType()` method and leave checking against LZ4 to the caller?

File: include/sas.h (hunk 1 above)
Comment (rkd-msw) [inline include/sas.h]: sas-client compiles with `-std=c++0x`, and (per the README) it needs to stay compatible with GCC 4.4 on RHEL, so I'll stick with the existing code.

File: include/sas.h [hunk 3]
```diff
@@ -114,42 +110,46 @@ class SAS
...
     static void destroy(void* compressor_ptr);
-
-    Compressor();
-    ~Compressor();
-
-    static const int WINDOW_BITS = 15;
-    static const int MEM_LEVEL = 9;
+    
+    static Compressor* get_zlib();
+    static Compressor* get_lz4();
```
Comment (mirw) [inline include/sas.h]: Shouldn't these methods live (as static methods) in LZ4Compressor and ZlibCompressor?

File: include/sas.h [hunk 4]
```diff
@@ -114,42 +110,46 @@ class SAS
...
     // Variables with which to store a compressor on a per-thread basis.
     static pthread_once_t _once;
-    static pthread_key_t _key;
-
-    z_stream _stream;
-    char _buffer[4096];
+    static pthread_key_t _zlib_key;
+    static pthread_key_t _lz4_key;
```
Comment (mirw) [inline include/sas.h]: ...and shouldn't these fields live there too?

File: include/sas.h [hunk 5]
```diff
@@ -114,42 +110,46 @@ class SAS
...
   };
 
   class Compressor
   {
   public:
-    static Compressor* get();
-
-    std::string compress(const std::string& s, const Profile* profile = NULL);
+    virtual std::string compress(const std::string& s, std::string dictionary) = 0;
+    static Compressor* get(const Profile* profile);
```
Comment (mirw) [inline include/sas.h]: Could the get method only depend on the CompressionType, not the full Profile?  (Would that be clearer?)

...and should CompressionType be nested within Compressor rather than Profile?

I don't feel too strongly, but that would feel more natural.

File: source/sas_compress.cpp [hunk 6]
```diff
@@ -41,19 +41,64 @@
...
+  ZlibCompressor();
+  ~ZlibCompressor();
+  std::string compress(const std::string& s, std::string dictionary);
+
+private:
+  static const int WINDOW_BITS = 15;
+  static const int MEM_LEVEL = 9;
+  // Variables with which to store a compressor on a per-thread basis.
+  static pthread_once_t _once;
+  static pthread_key_t _key;
```
Comment (mirw) [inline source/sas_compress.cpp]: With current design (of the `get_zlib` method being on the Compressor superclass), I'm not sure you need `_key` or `_once` here... although my recommendation is to move `get_zlib` onto ZlibCompressor and use these fields after all.

File: source/sas_compress.cpp [hunk 7]
```diff
@@ -101,18 +171,16 @@ SAS::Compressor::Compressor()
...
 /// Compressor destructor.  Terminates the zlib compressor.
-SAS::Compressor::~Compressor()
+ZlibCompressor::~ZlibCompressor()
 {
   deflateEnd(&_stream);
 }
 
 /// Compresses the specified string using the optional profile.
-std::string SAS::Compressor::compress(const std::string& s, const Profile* profile)
+std::string ZlibCompressor::compress(const std::string& s, std::string dictionary)
```
Comment (mirw) [inline source/sas_compress.cpp]: Why this change?  Isn't it still reasonable to pass the full Profile in, rather than pulling out the dictionary?

**Comment by mirw** [comment]
Looks good.  A few design tweaks suggested.

File: include/sas.h (hunk 1 above)
Comment (mirw) [inline include/sas.h]: Fine.

File: source/sas_compress.cpp (hunk 7 above)
Comment (rkd-msw) [inline source/sas_compress.cpp]: It is, but all we need is the dictionary, so it felt simpler just to pass in the piece of information we needed.

I think this is similar to your suggestion at https://github.com/Metaswitch/sas-client/pull/61#discussion_r47219168 - I'd like to be consistent, and either always extract the necessary piece of information from the Profile and pass that to the compressor, or always pass in the Profile and let the compressor do the extracting. I don't feel strongly - I'll do the former unless you feel strongly.

File: source/sas_compress.cpp (hunk 6 above)
Comment (rkd-msw) [inline source/sas_compress.cpp]: Agreed.

File: include/sas.h (hunk 5 above)
Comment (rkd-msw) [inline include/sas.h]: I'd prefer not to have CompressionType/Algorithm be nested within Compressor, just because I think the API user currently never has to see Compressor at the moment, and this would change that.

I'm happy to just pass in the `SAS::Profile::Algorithm` here, though.

**Comment by rkd-msw** [comment]
Thanks! All comments look good apart from the ones I've raised concerns about.

File: include/sas.h (hunk 5 above)
Comment (mirw) [inline include/sas.h]: Fair point.  Happy with that.

File: source/sas_compress.cpp (hunk 7 above)
Comment (mirw) [inline source/sas_compress.cpp]: I don't feel strongly.

**Comment by rkd-msw** [comment]
Rowena, can you please review this change from a SAS team perspective? It's already been reviewed within the Clearwater team.

You're also reviewing my design, which I've sent by email - let me know if you have any design feedback which I should apply before you do the code review.

**Comment by rkd-msw** [comment]
Markups made. I've checked that gzip compression still works with this code - I'm just about to confirm that LZ4 compression works and then I'll merge.
